### ISSUE TITLE
[BUG] Large (multiple parts) blobs might be handled incorrectly

### ISSUE THREAD

**Issue by paulstelian97** [thesofproject/sof#7081]
**Describe the bug**

> A clear and concise description of what the bug is.

Whenever attempting to switch out the FIR EQ configuration at runtime (while playing), an incorrect error message (Filter length 252 exceeds limitation of build) shows up in the logs and playback stops. On the i.MX platform it also leads to a DSP hang (possibly due to unrelated improper handling of XRUN conditions). The firmware logs were taken via etrace (because of the hang)

> What have you tried to diagnose or workaround this issue?
Please also read https://thesofproject.github.io/latest/contribute/process/bug-tracking.html for further information on submitting bugs.

I have attempted to change the configuration `COMP_LEGACY_INTERFACE` to `y` in order to attempt to bypass the Module Adapter (the logs are with this configuration set to `y`, which tells me I didn't actually manage to remove it from the causes).

**To Reproduce**
> Steps to reproduce the behavior: (e.g. list commands or actions used to reproduce the bug)

In one terminal, run

    aplay -Dhw:1,0 -t raw -f S32_LE -c 2 -r 48000 /dev/urandom

In another terminal, run various sof-ctl commands with the flat/mid/pass (small, single part blobs) and loudness (multi-part blob, which is the problem). After the execution of the multi-part blob, which succeeds, the next one leads to the incorrect message and XRUN (combined with DSP hang on i.MX).

    # These three commands succeed no matter how much you repeat or randomize them
    sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_flat.txt
    sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_mid.txt
    sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_pass.txt
    # This command succeeds once, but breaks something internally
    sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_loudness.txt
    # This command fails, no matter which of the four available files is given
    sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_flat.txt

**Reproduction Rate**
> How often does the issue happen ? i.e. 1/10 (once in ten attempts), 1/1000 or all the time.
Does the reproduction rate vary with any other configuration or user action, if so please describe and show the new reproduction rate.

It is reproducible 100% of the time. Bug as stated only affects GCC build, I did not test XCC build (but the failure is unexpected as the limitation is not actually hit)

**Expected behavior**
> A clear and concise description of what you expected to happen.

The last command in the example above should succeed and work normally, like before the loudness filter is applied.

**Impact**
> What impact does this issue have on your progress (e.g., annoyance, showstopper)

Annoyance (I'd say P3)

**Environment**
1) Branch name and commit hash of the 2 repositories: sof (firmware/topology) and linux (kernel driver).
    * Kernel: (internal, based on Linux 6.1.1 and the corresponding SOF branch). Issue does not seem kernel related.
    * SOF: 575deda444886ac028015ad9153162b36c00bd97 (5 commits behind imx-stable-v2.4, the commits there are not related to this issue)
2) Name of the topology file
    * Topology: sof-imx8-eq-fir-wm8960.tplg
3) Name of the platform(s) on which the bug is observed.
    * Platform: imx8 (i.MX8QM). I will trust @dbaluta that this issue also reproduces on the other i.MX platforms (which is reasonable, as there doesn't seem to be a root cause in platform specific drivers, whether IPC or otherwise)

**Screenshots or console output**
> If applicable, add a screenshot (drag-and-drop an image), or console logs
(cut-and-paste text and put a code fence (\`\`\`) before and after, to help
explain the issue.

> Please also include the relevant sections from the firmware log and kernel log in the report (and attach the full logs for complete reference). Kernel log is taken from *dmesg* and firmware log from *sof-logger*. See https://thesofproject.github.io/latest/developer_guides/debugability/logger/index.html

    [  1419163694.284585] (         463.906219) c0 dma-trace          ../audio/eq_fir/eq_fir.c:251  ERROR Filter length 252 exceeds limitation for build.
    [  1419163819.909580] (         125.624992) c0 dma-trace          ../audio/eq_fir/eq_fir.c:433  ERROR eq_fir_process(), failed FIR setup
    [  1419163941.940825] (         122.031242) c0 dma-trace          ......./module/generic.c:251  ERROR module_process() error -22: for comp 2
    [  1419164069.440820] (         127.499992) c0 dma-trace          ......./module_adapter.c:671  ERROR module_adapter_copy() error ffffffea: module processing failed
    [  1419164222.617897] (         153.177078) c0 dma-trace          ....../pipeline-stream.c:142  ERROR pipeline_copy(): ret = -22, start->comp.id = 6, dir = 1

**Additional notes**

I will also request that someone runs these commands (adjusted to a given platform) on an Intel system. The bug reproduces if an xrun occurs on the last command, no matter how it is handled.

**Comment by paulstelian97** [comment]
The relevant code seems to have been introduced in commit 99665561059b0c6f9d806772b4bddfdc0297f65c.

**Comment by dbaluta** [comment]
@singalsu @lgirdwood Can someone check if this also happens on Intel platforms?

What you need to do is: on a platform with EQ FIR run:
```
$ aplay
# while aplay is running
$ sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_flat.txt
$ sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_loudness.txt
$ sof-ctl -Dhw:1 -n 44 -s /unit_tests/sof/tools/ctl/eq_fir_mid.txt <--- after this aplay wil stop running
```

**Comment by singalsu** [comment]
I tested here with IPC3 build with topology tools/build_tools/topology/topology1/development/sof-hda-generic-4ch-loud.tplg with UP-extreme i11

```
$ amixer controls | grep fir
numid=33,iface=MIXER,name='EQFIR1.0 eqfir_coef_1'

$ sof-ctl -Dhw:0 -n 33 -s ctl3/eq_fir_flat.txt 
Control size is 4096.
Applying configuration "ctl3/eq_fir_flat.txt" into device hw:0 control numid=33.
4607827,0,56,50331648,0,0,0,0,56,65538,0,0,0,0,0,4294901764,0,0,0,0,16384,0
Success.
hdr: magic 0x00464f53
hdr: type 0
hdr: size 56 bytes
hdr: abi 3:0:0
4607827,0,56,50331648,0,0,0,0,56,65538,0,0,0,0,0,4294901764,0,0,0,0,16384,0

$ sof-ctl -Dhw:0 -n 33 -s ctl3/eq_fir_loudness.txt 
Control size is 4096.
Applying configuration "ctl3/eq_fir_loudness.txt" into device hw:0 control numid=33.
4607827,0,584,50413568,0,0,0,0,584,131076,0,0,0,0,65537,65537,4294901764,0,0,0,0,16384,0,252,0,0,0,0,65537,65537,65537,65537,131073,131074,131074,196611,196611,262147,262148,327685,393221,393222,458759,524296,589833,655370,720907,786444,851981,983054,1048591,1179665,1245202,1376276,1507350,1638424,1769498,1966109,2162719,2359330,2621478,2818089,3080237,3407922,3670070,3997755,4325440,4718661,5046347,5505105,5963863,6422622,6815845,7274604,7733362,8192121,8781953,9568396,10420376,11272357,12058802,12976318,14155982,15335649,16056559,16515324,16711960,24117551,70844611,3920495315,3920522431,70845139,24117443,16711983,16515352,16056572,15335663,14156001,12976334,12058814,11272370,10420389,9568408,8781964,8192129,7733369,7274610,6815852,6422629,5963870,5505111,5046353,4718667,4325445,3997760,3670075,3407926,3080242,2818093,2621481,2359334,2162722,1966111,1769501,1638426,1507352,1376278,1245204,1179666,1048593,983055,851982,786445,720908,655371,589834,524297,458760,393223,393222,327685,262149,262148,196611,196611,131075,131074,131074,65537,65537,65537,65537,1,0
Success.
hdr: magic 0x00464f53
hdr: type 0
hdr: size 584 bytes
hdr: abi 3:20:0
4607827,0,584,50413568,0,0,0,0,584,131076,0,0,0,0,65537,65537,4294901764,0,0,0,0,16384,0,252,0,0,0,0,65537,65537,65537,65537,131073,131074,131074,196611,196611,262147,262148,327685,393221,393222,458759,524296,589833,655370,720907,786444,851981,983054,1048591,1179665,1245202,1376276,1507350,1638424,1769498,1966109,2162719,2359330,2621478,2818089,3080237,3407922,3670070,3997755,4325440,4718661,5046347,5505105,5963863,6422622,6815845,7274604,7733362,8192121,8781953,9568396,10420376,11272357,12058802,12976318,14155982,15335649,16056559,16515324,16711960,24117551,70844611,3920495315,3920522431,70845139,24117443,16711983,16515352,16056572,15335663,14156001,12976334,12058814,11272370,10420389,9568408,8781964,8192129,7733369,7274610,6815852,6422629,5963870,5505111,5046353,4718667,4325445,3997760,3670075,3407926,3080242,2818093,2621481,2359334,2162722,1966111,1769501,1638426,1507352,1376278,1245204,1179666,1048593,983055,851982,786445,720908,655371,589834,524297,458760,393223,393222,327685,262149,262148,196611,196611,131075,131074,131074,65537,65537,65537,65537,1,0

$ sof-ctl -Dhw:0 -n 33 -s ctl3/eq_fir_mid.txt 
Control size is 4096.
Applying configuration "ctl3/eq_fir_mid.txt" into device hw:0 control numid=33.
4607827,0,128,50380800,0,0,0,0,128,65538,0,0,0,0,0,65576,0,0,0,0,1202154746,4222755303,3873958283,4024953600,4126012299,4176934804,4217764409,4250664050,4273864222,4288806716,2490349,6750285,8126582,7733371,6488174,4980824,3473472,2293804,1245211,1245197
Success.
hdr: magic 0x00464f53
hdr: type 0
hdr: size 128 bytes
hdr: abi 3:12:0
4607827,0,128,50380800,0,0,0,0,128,65538,0,0,0,0,0,65576,0,0,0,0,1202154746,4222755303,3873958283,4024953600,4126012299,4176934804,4217764409,4250664050,4273864222,4288806716,2490349,6750285,8126582,7733371,6488174,4980824,3473472,2293804,1245211,1245197

$ sof-ctl -Dhw:0 -n 33 -s ctl3/eq_fir_flat.txt 
Control size is 4096.
Applying configuration "ctl3/eq_fir_flat.txt" into device hw:0 control numid=33.
4607827,0,56,50331648,0,0,0,0,56,65538,0,0,0,0,0,4294901764,0,0,0,0,16384,0
Success.
hdr: magic 0x00464f53
hdr: type 0
hdr: size 56 bytes
hdr: abi 3:0:0
4607827,0,56,50331648,0,0,0,0,56,65538,0,0,0,0,0,4294901764,0,0,0,0,16384,0

```

I didn't see/hear any playback issues.

**Comment by paulstelian97** [comment]
Did you check for XRUNs that were handled? I believe they are handled better on Intel platforms than on ours.

**Comment by singalsu** [comment]
Is your build gcc or xcc? Gcc build is a lot more in risk for xruns because of high processing times.

**Comment by paulstelian97** [comment]
On my GCC build I get an error from FIR, -22 from the processing function, which triggers the xrun. Specifically I hit the FIR coefficients limit error (saying that I'm giving it 252 coefficients on one channel, which is nonsense). XCC build doesn't have that check compiled in the first place, and I don't know how this issue behaves _there_. Then DSP hangs from an IRQ storm (probably because the hardware XRUNs and I'm not handling it)

**Comment by singalsu** [comment]
Yep, that error was added to gracefully stop gcc build playback for too long filter for some (APL?) platform to handle it. I'll try here what gcc build does.

**Comment by singalsu** [comment]
Confirming that on gcc build the playback is not stopping with apply of eq_fir_loudness.txt. I get an  IPC timeout to kernel log and exit from trace so I can't see the error. I think I would need @kv2019i 's patch to see the error. So the handling of error from module adapter process seems to not work.

**Comment by kv2019i** [comment]
@singalsu Mtrace panic handling PR was merged to kernel last Friday, so you should now be able to get DSP panic dumps in mtrace without any extra patches (either to kernel or SOF FW).

**Comment by singalsu** [comment]
I was able to replicate the issue and get to trace "Filter length 252 exceeds limitation for build." and then get problems those appear and sound different in IPC3 and IPC4. But both failed to handle correctly processing component error propagation to stop playback without further issues.

**Comment by paulstelian97** [comment]
I have noticed something completely different. In commit c0b9b4057e4dc408c9af5c9695b9d052ab1d3702 some things are reordered in a weird way that makes the error from a blob to only show up when receiving the _next_ blob. The actual transfer is in fact fine. I'm currently trying to figure things out and I expect to send a patch to switch the order back around in a few hours.

Specifically, fir_init_coef being run before vs after the check (and the check being affected _by_ what is changed by fir_init_coef). I need to figure out how to best fix it, but I'm pretty sure it's now a bug in eq_fir rather than one in IPC/data_blob.

**Comment by dbaluta** [comment]
@singalsu @paulstelian97 I think you should sent a RFC patch with reordering so that people can comment. I'm not sure this change was intentional.


This is way I also ask people to separate patches into smaller chunks so that functional changes like the one brought by https://github.com/thesofproject/sof/commit/c0b9b4057e4dc408c9af5c9695b9d052ab1d3702 
to be well documented.

diff --git a/src/audio/eq_fir/eq_fir.c b/src/audio/eq_fir/eq_fir.c
index e10aff6cf..b9f33a7e8 100644
--- a/src/audio/eq_fir/eq_fir.c
+++ b/src/audio/eq_fir/eq_fir.c
@@ -312,6 +312,9 @@ static int eq_fir_init_coef(struct comp_dev *dev, struct sof_eq_fir_config *conf
                        return -EINVAL;
                }
 
+               fir_init_coef(&fir[i], eq);
+               comp_info(dev, "eq_fir_init_coef(), ch %d is set to response = %d", i, resp);
+
 #if defined FIR_MAX_LENGTH_BUILD_SPECIFIC
                if (fir[i].taps * nch > FIR_MAX_LENGTH_BUILD_SPECIFIC) {
                        comp_err(dev, "Filter length %d exceeds limitation for build.",
@@ -320,8 +323,6 @@ static int eq_fir_init_coef(struct comp_dev *dev, struct sof_eq_fir_config *conf
                }
 #endif
 
-               fir_init_coef(&fir[i], eq);
-               comp_info(dev, "eq_fir_init_coef(), ch %d is set to response = %d", i, resp);
        }
 
        return size_sum;

**Comment by paulstelian97** [comment]
I have reason to believe that the change to move fir_init_coef after the check is intentional and the correct fix is actually to change the way the check is being done. I will submit a patch like yours (except the comp_info is not copied) as a RFC.

**Comment by singalsu** [comment]
To me this looks like a pipeline / scheduler issue, not FIR. It can't stop gracefully the playback if one pipeline fails and is set to reset state.

**Comment by dbaluta** [comment]
I wonder how Intel handles this? I know that on I.MX when didn't really handled the errors on processing function and just let them be errros and reported on the linux kernel side.

But I guess on the processing function and looking at the general pipeline there can be some sort of error recovery.

**Comment by singalsu** [comment]
I was considering to set pass-through mode to be able to continue after illegal blob. But it could make the output a lot louder in some cases. As safe recover choice the EQ could switch to mute mode with filter with zero as gain when there's an error.

Or get all related pipelines to reset if there's an error at pipeline level. Though I don't know how to achieve that. What do you think?

**Comment by paulstelian97** [comment]
@singalsu I have a different idea, of simply rejecting invalid configuration as they're being received (rather than during the next process cycle) and using the old, already valid one. For an example see my new pull request #7129.

**Comment by lgirdwood** [comment]
@jsarha any comments ?

**Comment by paulstelian97** [comment]
The issue title is misleading, what I _thought_ caused the issue I was seeing. Really the main issue was what #7109 fixed (where the delayed appearance of the error made me think something is wrong in the infrastructure as opposed to inside the EQ FIR implementation itself).

**Comment by lgirdwood** [comment]
@paulstelian97 are you saying we can close this now ?

**Comment by dbaluta** [comment]
@lgirdwood we can close this once https://github.com/thesofproject/sof/pull/7129 is merged.

**Comment by paulstelian97** [comment]
Closing as #7129 got merged, and the original issue is resolved.
